damnit/backend/extraction_control.py

```diff
@@ -0,0 +1,247 @@
+            # Note: we put the run number first so that it's visible in
+            # squeue's default 11-character column for the JobName.
+            '--job-name', f"r{req.run}-p{req.proposal}-damnit",
+            '--wrap', shlex.join(req.python_cmd())
+        ]
+
+    def execute(self, req: ExtractionRequest):
+        """Run an extraction job in srun with live output"""
+        log_path = process_log_path(req.run, req.proposal, self.context_dir)
+        log.info("Processing output will be written to %s",
+                 log_path.relative_to(self.context_dir.absolute()))
+
+        # Duplicate output to the log file
+        with tee(log_path) as pipe:
+            subprocess.run(
+                self.srun_cmd(req), stdout=pipe, stderr=subprocess.STDOUT, check=True
+            )
+
+    def srun_cmd(self, req: ExtractionRequest):
+        return [
+            'srun', *self._resource_opts(req.cluster),
+            '--job-name', f"r{req.run}-p{req.proposal}-damnit",
+            *req.python_cmd()
+        ]
+
+    def _resource_opts(self, cluster: bool):
+        if cluster:
+            return self._slurm_cluster_opts()
+        else:
+            return self._slurm_shared_opts()
+
+    def _slurm_shared_opts(self):
+        return [
+            "--clusters", "solaris",
+            # Default 4 CPU cores & 25 GB memory, can be overridden
+            '--cpus-per-task', str(self.db.metameta.get('noncluster_cpus', '4')),
+            '--mem', self.db.metameta.get('noncluster_mem', '25G'),
+        ]
+
+    def _slurm_cluster_opts(self):
+        # Maxwell (dedicated node)
+        opts = [
+            "--clusters", "maxwell",
+            "--time", self.db.metameta.get("slurm_time", "02:00:00")
+        ]
+
+        if reservation := self.db.metameta.get('slurm_reservation', ''):
+            opts.extend(['--reservation', reservation])
+        else:
+            partition = self.db.metameta.get('slurm_partition', '') or default_slurm_partition()
+            opts.extend(['--partition', partition])
+
+        return opts
+
+
+def reprocess(runs, proposal=None, match=(), mock=False, watch=False):
+    """Called by the 'amore-proto reprocess' subcommand"""
+    submitter = ExtractionSubmitter(Path.cwd())
+    if proposal is None:
+        proposal = submitter.proposal
+
+    if runs == ['all']:
+        rows = submitter.db.conn.execute("SELECT proposal, run FROM runs").fetchall()
+
+        # Dictionary of proposal numbers to sets of available runs
+        available_runs = {}
+        # Lists of (proposal, run) tuples
+        props_runs = []
+        unavailable_runs = []
+
+        for proposal, run in rows:
+            if not mock and proposal not in available_runs:
+                available_runs[proposal] = proposal_runs(proposal)
+
+            if mock or run in available_runs[proposal]:
+                props_runs.append((proposal, run))
+            else:
+                unavailable_runs.append((proposal, run))
```

Comment:
```suggestion
        if mock:
            props_runs = rows
        else:
            for proposal, run in rows:
                if proposal not in available_runs:
                    available_runs[proposal] = proposal_runs(proposal)

                if run in available_runs[proposal]:
                    props_runs.append((proposal, run))
                else:
                    unavailable_runs.append((proposal, run))
```
To me this seems a bit cleaner. Although I'm not sure if `props_runs = rows` does the right thing or if rows needs to be looped on.

Reply:
Thanks, agreed that that seems cleaner. :+1: